File: hw/basurto/u3/hw25TemplateMethod/hw25TemplateMethod.py/model/caffeine_beverage.py
from abc import ABC, abstractmethod

class CaffeineBeverage(ABC):
    def prepare_recipe(self): # Template method
        self.boil_water()
        self.brew()
        self.pour_in_cup()
        if self.wants_condiments():
            self.add_condiments()

    def boil_water(self):
        print("Boiling water")

    @abstractmethod
    def brew(self):
        pass

    def pour_in_cup(self):
        print("Pouring into cup")

    @abstractmethod
    def add_condiments(self):
        pass

    def wants_condiments(self): # Hook operation
        return True

File: hw/alban/u3/hw25TemplateMethod/hw25TemplateMethod.js/model/CaffeineBeverage.js
class CaffeineBeverage {
    prepareRecipe() { // Template method
        this.boilWater();
        this.brew();
        this.pourInCup();
        if (this.wantsCondiments()) {
            this.addCondiments();
        }
    }

    boilWater() {
        console.log("Boiling water");
    }

    brew() { throw new Error("Must implement brew()"); }

    pourInCup() {
        console.log("Pouring into cup");
    }

    addCondiments() { throw new Error("Must implement addCondiments()"); }

    wantsCondiments() { return true; } // Hook operation
}
export default CaffeineBeverage;